src/batched-json.js

```diff
@@ -60,18 +60,20 @@ export function batchFetchJsonFor(
     // Throw user error if this element is performing URL substitutions
     // without the soon-to-be-required opt-in (#12498).
     if (opt_urlReplacement == UrlReplacementPolicy.OPT_IN) {
-      urlReplacements.collectUnwhitelistedVars(element).then(unwhitelisted => {
-        if (unwhitelisted.length > 0) {
+      return urlReplacements.collectUnwhitelistedVars(element).then(invalid => {
```

Comment:
Couldn't this be sync?

Also, isn't this really specific for `UrlReplacements`?

Reply:
Could be, but I kept it async so it can use `collectVars` internally which is async. I'll try massaging this to make it sync though.

It is, but fetching and URL replacements often have [tightly coupled business logic](https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md). `amp-list` and `amp-state` share this logic in particular.

Reply:
Made this sync which is much better, thanks.